plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemory.hs

```diff
@@ -262,17 +302,138 @@ instance ExMemoryUsage a => ExMemoryUsage [a] where
+addCostStreamGo :: CostStream -> CostStream -> CostStream
+addCostStreamGo (CostLast costL)        costsR = CostCons costL costsR
+addCostStreamGo (CostCons costL costsL) costsR = CostCons costL $ addCostStreamGo costsR costsL
+
+-- | Add two streams by interleaving their elements (as opposed to draining out one of the streams
+-- before starting to take elements from the other one). No particular reason to prefer
+-- interleaving over draining out one of the streams first.
+addCostStream :: CostStream -> CostStream -> CostStream
+addCostStream costsL0 costsR0 = case (costsL0, costsR0) of
+    (CostLast costL, CostLast costR) -> CostLast $ costL + costR
+    _                                -> addCostStreamGo costsL0 costsR0
+{-# INLINE addCostStream #-}
+
+-- See Note [Global local functions].
+-- Didn't attempt to optimize it.
+minCostStreamGo :: CostStream -> CostStream -> CostStream
+minCostStreamGo costsL costsR =
+    -- Peel off a cost from each of the streams, if there's any, compare the two costs, emit
+    -- the minimum cost to the outside and recurse. If the two elements aren't equal, then we put
+    -- the difference between them back to the stream that had the greatest cost (thus subtracting
+    -- the minimum cost from the stream -- since we just accounted for it by lazily emitting it to
+    -- the outside). Proceed until one of the streams is drained out.
+    let (!costL, !mayCostsL') = unconsCost costsL
+        (!costR, !mayCostsR') = unconsCost costsR
+        (!costMin, !mayCostsL'', !mayCostsR'') = case costL `compare` costR of
+            LT -> (costL, mayCostsL', pure $ reconsCost (costR - costL) mayCostsR')
+            EQ -> (costL, mayCostsL', mayCostsR')
+            GT -> (costR, pure $ reconsCost (costL - costR) mayCostsL', mayCostsR')
+    in reconsCost costMin $ minCostStreamGo <$> mayCostsL'' <*> mayCostsR''
+
+-- | Calculate the minimum of two 'CostStream's. May return a stream that is longer than either of
+-- the two (but not more than twice).
+minCostStream :: CostStream -> CostStream -> CostStream
+minCostStream costsL0 costsR0 = case (costsL0, costsR0) of
+    (CostLast costL, CostLast costR) -> CostLast $ min costL costR
+    _                                -> minCostStreamGo costsL0 costsR0
+{-# INLINE minCostStream #-}
+
+-- See Note [Global local functions].
+-- Unboxing @Int#@s manually, because no matter what I tried, GHC just wouldn't unbox an argument
+-- of a function passed as an accumulator.
+flattenCostRoseGo :: Int# -> [CostRose] -> (Int# -> CostStream) -> CostStream
+flattenCostRoseGo i0 []                                    k = k i0
+-- I cannot imagine this tilda making a difference in any even remotely realistic scenario, but
+-- better safe than sorry.
+flattenCostRoseGo i0 (~(CostRose cost1 forest1) : forest2) k =
+    CostCons (coerce $ I# i0) $
+        case coerce cost1 of
+            I# i1 ->
+                flattenCostRoseGo i1 forest1 $ \i2 ->
```

Comment:
As far as I can tell this is amortized O(1) for producing each element, but could take linear time for some elements (e.g., the last one). Is this optimal?

Reply:
Hm, I'm not seeing any linear time in there, could you explain? Linear memory in the _depth_ of tree -- sure, but I don't think we can do better than that and it's conservative too (it's just that previously the linearity-of-memory-consumption was in the `ExMemoryUsage Data` instance rather than here).

So `flattenCostRoseGo` always immediately emits an element when the forest is non-empty. Then it grows the continuation (constant-time too) and recurses into the new node. If the node is non-empty too, then we get another element in constant time, grow the continuation in constant time and recurse again. If the next node contains an empty forest, then we don't emit an element and jump to the continuation, which in constant time to reduces to a `flattenCostRoseGo` call (i.e. without forcing the entirety of the so-far-accumulated continuation, just by reducing the outermost lambda), which again emits a cost immediately.

Thus, it seems to me that every cost is emitted in constant time, but maybe I'm misunderstanding something?

Reply:
```haskell
flattenCostRose (CostRose 1 [r1]) where
  r1 = CostRose 2 [r2]
  r2 = CostRose 3 [r3]
  r3 = CostRose 4 []
```

In this case it will reduce to

```haskell
CostCons 1 $ CostCons 2 $ CostCons 3 $ <continuation that produces CostLast 4 in 4 steps>
```

The continuation is essentially
```haskell
\x1 -> flattenCostRoseGo x1 [] $
  \x2 -> flattenCostRoseGo x2 [] $
    \x3 -> flattenCostRoseGo x3 [] $
      \x4 -> CostLast x4
```

So it seems producing `4` takes linear time.

Reply:
OK, I was indeed misunderstanding, thanks a lot for the clarification and for catching. I'll think if it's a problem (releasing a linear amount of memory is a linear operation anyway, so maybe we can wait for the final cost as well, but not sure) and what to do about it.

Reply:
> So it seems producing 4 takes linear time.

(linear in the depth of the tree, just like memory consumption)

Reply:
If it's amortized O(1) then it's not a problem. I think the following is O(1) worst case, but I wouldn't be surprised if it's slower:
```haskell
flattenCostRose :: CostRose -> CostStream
flattenCostRose (CostRose cost []) = CostLast cost
flattenCostRose (CostRose cost (r:rs)) = CostCons cost (go r rs)

go :: CostRose -> [CostRose] -> CostStream
go (CostRose cost []) [] = CostLast cost
go (CostRose cost []) (r:rs) = CostCons cost (go r rs)
go (CostRose cost (r:rs)) rs' = CostCons cost (go r (rs ++ rs'))
```

Reply:
Your code doesn't look much different to me, but perhaps I'm misunderstanding again. E.g.

```
go [CostRose cost1 [CostRose cost2 [CostRose cost3 []]]] rs' ~>

CostRose cost1 $ go [CostRose cost2 [CostRose cost3 []]] ([] ++ rs') ~>

CostRose cost1 $ CostRose cost2 $ go [CostRose cost3 []] ([] ++ ([] ++ rs'))
```

It seems like for long chain of one-element forest we're going to end up with `[] ++ ([] ++ ([] ++ <...>))`, which is effectively the same, except we also pay for all the overhead of list concatenation (it's lazy, so costs are amortized, but they're still there).

> If it's amortized O(1) then it's not a problem. 

Yeah, and I feel like special-casing single-element forests could help us level the cost of retrieving the next element, but at the expense of introducing an additional cost for every single node, which wouldn't pay off, I think. Still, I'm going to play with it.

Reply:
How about this:

```haskell
flattenCostRoseGo :: Int# -> CostRose -> [CostRose] -> (Int# -> CostStream) -> CostStream
flattenCostRoseGo i0 rose1 forest2 k =
    CostCons (coerce $ I# i0) $
        let !(CostRose cost1 forest1) = rose1
            !(I# i1) = coerce cost1
        in case forest1 of
            [] -> case forest2 of
                []                -> k i1
                rose2' : forest2' -> flattenCostRoseGo i1 rose2' forest2' k
            rose1' : forest1' -> case forest2 of
                []                -> flattenCostRoseGo i1 rose1' forest1' k
                rose2' : forest2' ->
                    flattenCostRoseGo i1 rose1' forest1' $ \i2 ->
                        flattenCostRoseGo i2 rose2' forest2' k

flattenCostRose :: CostRose -> CostStream
flattenCostRose (CostRose cost [])              = CostLast cost
flattenCostRose (CostRose cost (rose : forest)) =
    case coerce cost of
        I# i -> flattenCostRoseGo i rose forest $ \iz -> coerce CostLast $ I# iz
{-# INLINE flattenCostRose #-}
```

We essentially need to prove by induction that both `flattenCostRoseGo` and its continuation argument always produce an element in constant time. For `flattenCostRoseGo` it's entirely obvious and for the continuation we have several cases:

1. the initial case with `CostLast` -- obviously constant-time
2. the case when either of `forest1` or `forest2` is empty (but not both at the same time) -- we pass the continuation unchanged and hence don't change any of its behavior
3. the case when both `forest1` and `forest2` are non-empty -- we do grow the continuation, but by induction hypothesis `flattenCostRoseGo` produces an element in constant time and so is the new continuation 

I.e. we break the previous chains of empty forest handling by matching on `forest2` _before_ recursing in order not to grow the continuation with a fancy form of `id`. And we also don't seem to do any extra work overall, since that matching is useful (now that `flattenCostRoseGo` receives both a node and a remaining forest), albeit slightly premature (we may run out of budget before getting to accounting for costs from `forest2`, but we don't care much about optimizing the unhappy path, since it's when the smart contracts fails anyway).

Reply:
> It seems like for long chain of one-element forest we're going to end up with [] ++ ([] ++ ([] ++ <...>)), which is effectively the same

Yeah, I think you are right.

> What do you think?

Yes I think this is indeed O(1) worst case. However, I don't quite see the benefit of CPS, because in my non-CPS version, you can also match on the `rs` and avoid concatenation if it is empty.

Reply:
> However, I don't quite see the benefit of CPS, because in my non-CPS version, you can also match on the `rs` and avoid concatenation if it is empty.

Yours is different. In mine we use the matching not only to check that the list is not empty, but also to retrieve its head to use it later, so in any case the matching is useful. In your case it's only useful when the list is empty, provided I got the code right:

```haskell
flattenCostRoseGo :: CostRose -> [CostRose] -> CostStream
flattenCostRoseGo (CostRose cost1 forest1) forest2 =
    case forest1 of
        [] -> case forest2 of
            []                -> CostLast cost1
            rose2' : forest2' -> CostCons cost1 $ flattenCostRoseGo rose2' forest2'
        rose1' : forest1' ->
            CostCons cost1 $ case forest1' of
                [] -> flattenCostRoseGo rose1' forest2
                _  -> flattenCostRoseGo rose1' $ forest1' ++ forest2

flattenCostRose :: CostRose -> CostStream
flattenCostRose (CostRose cost [])              = CostLast cost
flattenCostRose (CostRose cost (rose : forest)) = CostCons cost $ flattenCostRoseGo rose forest
{-# INLINE flattenCostRose #-}
```

That is one overhead, the other one is `++`. But maybe both of these overheads are negligible compared to the overhead of passing a continuation around.

I'll run the validation benchmarks with your version. If the results are the same, I'll pick it as it's much simpler than the continuation one.

Although I also wanted to try a different definition of `CostRose`.